### PR TITLE
Onnx support

### PR DESCRIPTION
**Pull request by aicelen**
As the title already says this PR changes homr's inference backend from torch to onnx. I did not integrate the onnx backend into the colab notebook therefore it might break.

## Code changes
The converting code is located in `homr/convert_onnx` entry point is `main.py`.
The inference code is still in `homr/segmentation `and `homr/transformer`.
The inference entry point of tromr is still in `staff2score.py`

## Features & Advantages

- Full GPU support 
- Faster inference: tromr is around 2x faster (6 core CPU); segmenting on a T4 (colab) at `batch_size=8` is around 20ms per image. This means segmenting one sheet (3000x4000 pixel) can be done in less than 3 seconds.
- No torch required for inference (large download size)
- Easier crossplatform integration (for example Android)
- I did not notice a loss in quality

## Difficulties & Future Ideas

- Tromr encoder is hard to quantize
- Segnet could be quantized using a static quant (requires dataset) or dynamic with quint (worse performance on x86-64; around 50% worse)

## Notes
- In tromr's decoder I changed torch.multinominal() to np.argmax(). I don't think it really matters because the output values were (from my experience) far apart.

## Technical details about converting tromr to onnx
I seperated the enocder and decoder. The encoder was pretty easy to convert. 
On the other side it was way more difficult to convert the decoder since the inference code in `ScoreTransformerWrapper()` could not be converted to onnx. Therefore I only converted the `ScoreDecoder()` and replicated the inference code of  `ScoreTransformerWrapper()` in `decoder_inference.py`. I also quantized the decoder to qint8.

### CODE REVIEW

**Comment by liebharc** [comment]
Thanks a lot @aicelen, I'm just about to leave and will travel for a couple of days. I will look into the changes as soon as I can!

**Comment by aicelen** [comment]
Hi
thank you for your feedback! I'm currently using version 160 of the transformer. Updating it to 180 breaks it. I'll have a look into that but it might take some time. I'll comment here when it's finished. 
By the way shall I include the hex values of the pytorch model names also in the names of the onnx models (or is for example segnet_155.onnx enough)?

**Comment by aicelen** [comment]
Here's a quick update.
- converting for v188 of transformer working
- onnx models include version+hash
- downloading models works (it's currently downloading from my fork; maybe you move them here once this is merged)
- config files are in sync again

I only need to move the architecture and the torch inference code to the training folder. 
I think it would be better to make a training/transformer/arch and a training/transformer/inference folder so there aren't too many files in training/transformer. Is that also fine?

**Comment by liebharc** [comment]
Yes subfolders under training transformer are a good idea.

File: training/music_xml.py [hunk 1]
```diff
@@ -359,7 +359,9 @@ def music_xml_string_to_semantic(content: str) -> list[list[str]]:
 
 
 def music_xml_to_semantic(file_path: str) -> list[list[str]]:
-    with open(file_path) as file:
+    # using rb fixes the following error I got (windows)
```
[the remark this answers is not in the document]
Comment (aicelen) [reply training/music_xml.py]: Improved the comment in the latest commit

File: training/inference_torch/inference_segnet.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import cv2
+import numpy as np
+from PIL import Image
+from training.architecture.segmentation import config
+from training.architecture.segmentation.inference import inference
+from homr.type_definitions import NDArray
+from homr.color_adjust import color_adjust
+
+
+def generate_pred(image: NDArray, ouput_path, num_classes) -> tuple[NDArray, NDArray, NDArray, NDArray, NDArray]:
```
[the remark this answers is not in the document]
Comment (aicelen) [reply training/inference_torch/inference_segnet.py]: Ok. I just deleted the whole inference_torch folder and moved the testing script for the transformer into training/architecture/staff2score.py
It could be useful for testing on a single image.

File: training/convert_onnx/segmentation/model.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
+import segmentation_models_pytorch as smp
+import torch
+from torch.optim import lr_scheduler
+
+
+class CamVidModel(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (aicelen) [reply training/convert_onnx/segmentation/model.py]: Leftover from moving the converting script into the homr branch...
Fixed it with the latest commit.

File: training/architecture/transformer/tokenizer_lift.json [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{
+	"version": "1.0",
```
[the remark this answers is not in the document]
Comment (aicelen) [reply training/architecture/transformer/tokenizer_lift.json]: Fixed it.

File: homr/transformer/staff2score.py [hunk 5]
```diff
@@ -1,115 +1,52 @@
...
-    ) -> list[TransformerChord]:
-        return self.model.generate(
-            imgs_tensor,
-            debug=debug,
-        )
+        t0 = perf_counter()
 
+        # Create special tokens
+        start_token = np.full((len(x), 1), self.config.bos_token, dtype=np.int64)
+        nonote_token = np.full((len(x), 1), self.config.nonote_token, dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (aicelen) [reply homr/transformer/staff2score.py:36]: Yes thank you for pointing that out. It actually fixed an error on my test_image

File: training/architecture/segmentation/inference.py [hunk 6]
```diff
@@ -1,83 +1,83 @@
...
+    step_size: int = 320
+) -> NDArray:
+    if "segnet" in model_path:
+        model = create_segnet(skip_weights_download=True)
+    elif "unet" in model_path:
+        model = create_unet(skip_weights_download=True)
+    else:
+        raise ValueError("Unknown model type: " + model_path)
+    model.load_state_dict(torch.load(model_path, weights_only=True), strict=False)
+    images = split_into_patches(image, win_size=320, step_size=step_size)
```
[the remark this answers is not in the document]
Comment (aicelen) [reply training/architecture/segmentation/inference.py]: We dont need the inference code for the segnet so i just deleted the file

**Review by aicelen — COMMENTED** [review]
I fixed the easy things. Later I'll look into debug and caching.

**Comment by aicelen** [comment]
I added all your requested changes except the debugging (and the argument for converting). 
Reintorducing the debugging mechanics is more complicated than I expected because I need to change things with the onnx model itself. I wondered if we really need the debugging mechanics in the onnx model because the debugging could also be done with the torch model. Of course that would limit the capabilities of debugging if the model breaks during converting it to onnx.
If the debugging mechanics are really needed I'll look into this later since I'm on a trip next week.
